---
title: "Random data ROC"
author: "Peter Hettegger"
date: "`r Sys.Date()`"
output:
  pdf_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(digits = 4)
```


This document provides a simulation example for evaluating ROC AUC values for:

1. batch effect correction with `ComBat`.
2. batch effect correction with `ComBat` and correction of p-values with 1000 random rotations.

We assume a sample design of 3 `batches` containing samples of two `groups` (control/case), 30 samples in total. We assume to have 5000 `features` or genes, where the first 1000 features are differentially expressed.


```{r message=FALSE}
library(randRotation)
library(pROC)
library(GGally)
set.seed(0)
```


```{r message=FALSE}
pdata <- data.frame(batch = as.factor(rep(c(1:3), c(10,10,10))),
                    group = as.factor(rep(c("Control", "Case"), c(5,5))))
with(pdata, table(batch, group))
```

We generate random normal data for 5000 features
```{r message=FALSE}
features <- 5000

# Matrix with random gene expression data
edata <- matrix(rnorm(features * nrow(pdata)), features)
rownames(edata) <- paste("feature", 1:nrow(edata))
```

and add a small artificial covariate effect to the first 1000 features:

```{r message=FALSE}
######## add covariate effect for features 1:1000 ###########
feat.sign <- 1:1000
edata[feat.sign,] <- t(t(edata[feat.sign,]) + as.numeric(pdata$group) - 1.5)
#############################################################
```


We now perform analysis (1) with batch effect correction by `ComBat`, but without random rotation:
```{r message=FALSE}
mod1 <- model.matrix(~group, pdata)

##### ComBat
edata.combat <- sva::ComBat(edata, pdata$batch, mod1)
fit1 <- limma::lmFit(edata.combat, mod1)
fit1 <- limma::eBayes(fit1)
tt1  <- limma::topTable(fit1, 2, Inf, sort.by = "none", adjust.method = "none")
p.com <- tt1$P.Value

hist(p.com)
```

Now we perform the same analysis, but with p-value correction by random rotation (analysis (2)):

```{r message=FALSE}
###### with random rotation

rr1 <- initBatchRandrot(edata, mod1, 2, pdata$batch)

statistic <- function(Y, mod, batch){
    Y <- sva::ComBat(Y, batch, mod)
    fit1 <- limma::lmFit(Y, mod)
    fit1 <- limma::eBayes(fit1)
    abs(fit1$t[,2])
}

rs1 <- rotateStat(rr1, 1000, statistic, mod1, pdata$batch, parallel = TRUE)
p.rot <- pFdr(rs1, pooled = TRUE)[,1]

hist(p.rot)
```

The following code generates the ROC curves for both analyses:

```{r message=FALSE}
r.com <- roc(cases = p.com[feat.sign], controls = p.com[-feat.sign])
r.com
r.rot <- roc(cases = p.rot[feat.sign], controls = p.rot[-feat.sign])
r.rot
plot(r.com, lty = 2, lwd = 2)
lines(r.rot, col = 2, lty = 4, lwd = 2)

legend("bottomright", legend = c("ComBat", "ComBat - 1000 rotations"),
       lty = c(2,4), col = 1:2, lwd = 2)
```


```{r}
df1 <- data.frame(p.com, p.rot)
colnames(df1) <- c("p ComBat", "p ComBat - 1000 rot.")

ggally_points(df1, aes(x = `p ComBat`, y = `p ComBat - 1000 rot.`), size = 0.7)+
    scale_y_log10()+
    scale_x_log10()+
    geom_abline(slope = 1, intercept = 0, lty = 2, lwd = 0.5, col = "red")
```



# Session info

```{r}
sessionInfo()
```




